### PR TITLE
Rewrite BN backwards to be 2 reduction approach, not 4.

### PR DESCRIPTION
**Pull request by csarofeen**
Don't merge before: https://github.com/csarofeen/pytorch/pull/1209 had to fix expr sorting and loop nest generation for this.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 1]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
-
-    auto c1 = mul(grad_x_hat, x_hat);
-    auto c2 = sum(c1, reduction_axes);
-    auto bcast_c2 = broadcast(c2, broadcast_mask);
-    auto c3 = mul(x_hat, bcast_c2);
+  mean = broadcast(mean, broadcast_mask);
+
+  TensorView* weight_val;
+  if (weight == nullptr) {
+    weight_val = TensorViewBuilder()
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:419]: The input tensor may not be only 4 dimensions. There is also 1D and 3D batch norm.
Maybe we should create a `ones_like` function for this.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 2]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
+                     .shape({1, 1, 1, 1})
+                     .build();
+    new UnaryOp(
+        UnaryOpType::Set, weight_val->as<Val>(), (new Double(1.0))->as<Val>());
+  } else {
+    weight_val = broadcast(weight, broadcast_mask);
+  }
 
-    auto inner = sub(sub(a, bcast_b), c3);
+  auto norm = div(new Double(1), num_features);
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: Use `reciprocal(num_features)` for `UnaryOpType::Reciprocal`

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 3]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
-  TensorView* db = nullptr;
+  auto mean = save_mean;
+  auto invstd = save_invstd;
   if (kTraining) {
     TORCH_INTERNAL_ASSERT(
         save_mean != nullptr && save_invstd != nullptr,
         "When training=True, save_mean and save_invstd are required.");
+  } else {
+    mean = running_mean;
+    invstd = div(new Double(1.0), sqrt(add(running_var, eps)));
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: Use `rsqrt(add(running_var, eps))` for `UnaryOpType::Rsqrt`

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 4]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
+  } else {
+    weight_val = broadcast(weight, broadcast_mask);
+  }
 
-    auto inner = sub(sub(a, bcast_b), c3);
+  auto norm = div(new Double(1), num_features);
 
-    auto reciprocal_size = unaryOp(UnaryOpType::Reciprocal, num_features);
+  auto grad_output_sum = sum(grad_output, reduction_axes);
+  auto dot_p = sum(mul(grad_output, sub(input, mean)), reduction_axes);
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp:433]: You can reuse `sub(input, mean)` in `dot_p` and `proj`

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 5]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
-    }
+  auto grad_mean = broadcast(mul(grad_output_sum, norm), broadcast_mask);
+  auto proj_scale =
+      broadcast(mul(mul(dot_p, norm), mul(invstd, invstd)), broadcast_mask);
+  auto grad_scale = mul(broadcast(invstd, broadcast_mask), weight_val);
 
-    if (output_mask[1]) {
-      dw = sum(mul(dy, x_hat), reduction_axes);
-    }
+  TensorView* grad_input;
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: `TensorView* grad_input = nullptr;`

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 6]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
...
-    auto b = sum(grad_x_hat, reduction_axes);
-    auto bcast_b = broadcast(b, broadcast_mask);
-
-    auto c1 = mul(grad_x_hat, x_hat);
-    auto c2 = sum(c1, reduction_axes);
-    auto bcast_c2 = broadcast(c2, broadcast_mask);
-    auto c3 = mul(x_hat, bcast_c2);
+  mean = broadcast(mean, broadcast_mask);
+
+  TensorView* weight_val;
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: `TensorView* weight_val = nullptr;`

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp [hunk 7]
```diff
@@ -379,90 +379,83 @@ BackwardNormResult batch_norm_backward(
   // N = reduction = B * H * W * D
   // weight = bias = (C) tensor
   const size_t kNumberOfDims =
-      TensorDomain::noReductions(x->getRootDomain()).size();
-  // channels last format means C dimension is at axis kNumberOfDims-1 at x / dy
+      TensorDomain::noReductions(input->getRootDomain()).size();
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/ops/normalization.cpp]: Should this be `getMaybeRFactorDomain()`?

**Review by rdspring1 — APPROVED** [review]
LGTM. The math looks fine and I am assuming it passes `grad_check` or all the tests.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ops/normalization.cpp:419]: Good catch, thanks.

File: torch/csrc/jit/codegen/cuda/ops/normalization.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ops/normalization.cpp:433]: I'm going to leave this for now as it will just shift the persistent buffer backwards when persistent, and when segmented we will recompute this anyways.
